### PR TITLE
add sharedvpc support for lb subnet

### LINKED ISSUE #1279: Internal Load Balancer doesn't support SharedVPC
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

Using the following load balancer configuration:

```yaml
spec:
  loadBalancer:
    internalLoadBalancer: ${NAME_OF_SHAREDVPC_SUBNET}
  loadBalancerType: "INTERNAL"
```

CAPG-controller-manager returns the following error:

```
E0701 14:43:51.301504       1 reconcile.go:525] "Error getting subnet for Internal Load Balancer" err="googleapi: Error 404: The resource 'projects/CURRENT_PROJECT/regions/us-central1/subnetworks/SHAREDVPC_SUBNET_NAME' was not found, notFound" controller="gcpcluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="GCPCluster" GCPCluster="NAMESPACE/CLUSTER_NAME" namespace="NAMESPACE" name="CLUSTER_NAME" reconcileID="ID_HASH"
```

**What did you expect to happen:**
The GCPCluster should use `spec.network.hostProject` to check `spec.loadBalancer.internalLoadBalancer.subnet`.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api version: v1.7.0
- Rancher: v2.7.8 
- Kubernetes version: (use `kubectl version`): 
```
kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.13", GitCommit:"49433308be5b958856b6949df02b716e0a7cf0a3", GitTreeState:"clean", BuildDate:"2023-04-12T12:15:50Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.12+rke2r1", GitCommit:"ef70d260f3d036fc22b30538576bbf6b36329995", GitTreeState:"clean", BuildDate:"2023-03-17T18:01:17Z", GoVersion:"go1.19.7 X:boringcrypto", Compiler:"gc", Platform:"linux/amd64"}
```
- OS (e.g. from `/etc/os-release`): Rocky Linux 9

### PR DESCRIPTION
**Pull request by swarren83**
**What type of PR is this?**

**What this PR does / why we need it**:
Add support for SharedVPC when specifying `spec.loadBalancer.internalLoadBalancer.subnet`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1279

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [x] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cloud/services/compute/loadbalancers/service.go [hunk 1]
```diff
@@ -113,6 +118,6 @@ func New(scope Scope) *Service {
 		regionalhealthchecks:    scope.Cloud().RegionHealthChecks(),
 		instancegroups:          scope.Cloud().InstanceGroups(),
```
Comment (barbacbd) [inline cloud/services/compute/loadbalancers/service.go:119]:
```suggestion
		instancegroups:          cloudScope.InstanceGroups(),
```

File: cloud/services/compute/loadbalancers/service.go [hunk 2]
```diff
@@ -113,6 +118,6 @@ func New(scope Scope) *Service {
...
 		targettcpproxies:        scope.Cloud().TargetTcpProxies(),
```
Comment (barbacbd) [inline cloud/services/compute/loadbalancers/service.go:120]: I believe that the health checks should be created or found in the VPC Network Project. I am unsure, but the Target TCP Proxies may be too.

File: cloud/services/compute/loadbalancers/service.go [hunk 3]
```diff
@@ -113,6 +118,6 @@ func New(scope Scope) *Service {
 		regionalhealthchecks:    scope.Cloud().RegionHealthChecks(),
```
Comment (barbacbd) [inline cloud/services/compute/loadbalancers/service.go:118]:
```suggestion
		regionalhealthchecks:    cloudScope.RegionHealthChecks(),
```

File: cloud/services/compute/loadbalancers/service.go (hunk 2 above)
Comment (swarren83) [reply cloud/services/compute/loadbalancers/service.go]: @barbacbd I updated to include `regionalHealthChecks` and `targetTCPProxies`.  I also updated the tests.  I finished signing the CLA and got the confirmation emails, but this PR is still failing on that step.

**Comment by swarren83** [comment]
@cpanato I signed the CLA yesterday and again today.  I have received email confirmation both times.  It states "This is a notification email from EasyCLA regarding the project OpenTelemetry."  According to the [docs](https://github.com/kubernetes/community/blob/master/CLA.md) it doesn't appear that EasyCLA is sending me to the correct agreement.  Is there another way to sign the CLA?

**Comment by swarren83** [comment]
I have opened a support case with LFX Support regarding EasyCLA issues I am experiencing.

**Comment by swarren83** [comment]
The EasyCLA issue was that I signed with an email address that didn't match the one used to sign the CLA.

**Comment by swarren83** [comment]
@cpanato I need to wait on merging this.  I need to work on switch to the Corporate CLA.

**Comment by swarren83** [comment]
@cpanato Ready to proceed if this lgtm.

**Review by cpanato — COMMENTED** [review]
/assign @damdo 

/approve

**Comment by swarren83** [comment]
@barbacbd I have reverted the changes to scope regional health checks and target tcp proxies.

**Comment by barbacbd** [comment]
> @barbacbd I have reverted the changes to scope regional health checks and target tcp proxies.

@swarren83 I would probably rebase it so that there is only one commit. Otherwise LGTM. @cpanato PTAL Thank you!

**Comment by barbacbd** [comment]
> @barbacbd I rebased to a single commit as requested. Do we need anything else to get this ready to merge?

I am good with this change. I do not have LGTM or Approval privileges in this repo. @damdo or @cpanato can you take a look please?

**Review by damdo — COMMENTED** [review]
Thanks for your PR @swarren83 

/lgtm
/unhold

**Pull request merged**